report/topology.go

```diff
@@ -195,6 +215,100 @@ func (c Counters) Copy() Counters {
 	return result
 }
 
+// Sets is a string->set-of-strings map.
+type Sets map[string]StringSet
+
+// Merge merges two sets maps into a fresh set, performing set-union merges as
+// appropriate.
+func (s Sets) Merge(other Sets) Sets {
+	result := s.Copy()
+	for k, v := range other {
+		result[k] = result[k].Merge(v)
+	}
+	return result
+}
+
+// Copy returns a value copy of the sets map.
+func (s Sets) Copy() Sets {
+	result := Sets{}
+	for k, v := range s {
+		result[k] = v.Copy()
+	}
+	return result
+}
+
+// StringSet is a sorted set of unique strings. Clients must use the Add
+// method to add strings.
+type StringSet []string
+
+// MakeStringSet makes a new StringSet with the given strings.
+func MakeStringSet(strs ...string) StringSet {
```

Comment:
Nice catch! I'll fix this first thing.

On Wednesday, 28 October 2015, Paul Bellamy <notifications@github.com>
wrote:

> In report/topology.go
> <https://github.com/weaveworks/scope/pull/522#discussion_r43277529>:
>
> > +
> > +// Copy returns a value copy of the sets map.
> > +func (s Sets) Copy() Sets {
> > +	result := Sets{}
> > +	for k, v := range s {
> > +		result[k] = v.Copy()
> > +	}
> > +	return result
> > +}
> > +
> > +// StringSet is a sorted set of unique strings. Clients must use the Add
> > +// method to add strings.
> > +type StringSet []string
> > +
> > +// MakeStringSet makes a new StringSet with the given strings.
> > +func MakeStringSet(strs ...string) StringSet {
>
> This function modifies the input, which could cause gotchas later on.
>
> e.g.
>
> a := []string{"b", "a", "c"}MakeStringSet(a...)
> fmt.Println(a) // [a b c]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/weaveworks/scope/pull/522/files#r43277529>.
>